File: src/main/java/bad/robot/radiate/teamcity/SanitisedException.java
package bad.robot.radiate.teamcity;

import static org.apache.commons.lang3.ClassUtils.getShortClassName;
import static org.apache.commons.lang3.StringUtils.*;
import static org.apache.commons.lang3.exception.ExceptionUtils.getRootCause;

public class SanitisedException {

    private final Exception underlying;

    public SanitisedException(Exception underlying) {
        this.underlying = underlying;
    }

    public String getMessage() {
        if (getRootCause(underlying) != null)
            return removeClassPreamble(getRootCause(underlying));
        return underlying.getMessage();
    }

    private static String removeClassPreamble(Throwable exception) {
        return getExpandedClassShortName(exception.getClass()) + ": " + defaultString(exception.getMessage());
    }

    public static String getExpandedClassShortName(Class type) {
        String name = getShortClassName(type);
        return join(splitByCharacterTypeCamelCase(name), " ");
    }
}

File: src/main/scala/bad/robot/radiate/teamcity/SanitisedException.scala
package bad.robot.radiate.teamcity

import org.apache.commons.lang3.ClassUtils.getShortClassName
import org.apache.commons.lang3.StringUtils._
import org.apache.commons.lang3.exception.ExceptionUtils.getRootCause

object SanitisedException {
  private def removeClassPreamble(exception: Throwable): String = {
    getExpandedClassShortName(exception.getClass) + ": " + defaultString(exception.getMessage)
  }

  def getExpandedClassShortName(`type`: Class[_]): String = {
    val name = getShortClassName(`type`)
    join(splitByCharacterTypeCamelCase(name), " ")
  }
}

class SanitisedException(underlying: Exception) {

  def getMessage: String = {
    if (getRootCause(underlying) != null) SanitisedException.removeClassPreamble(getRootCause(underlying))
    else underlying.getMessage
  }
}